ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderForInsertOnly.java

```diff
@@ -0,0 +1,274 @@
+    }
+  }
+
+  private static Set<String> getHiveMetastoreConstants() {
+    Set<String> result = new HashSet<>();
+    for (Field f : hive_metastoreConstants.class.getDeclaredFields()) {
+      if (!Modifier.isFinal(f.getModifiers())) {
+        continue;
+      }
+      if (!String.class.equals(f.getType())) {
+        continue;
+      }
+      f.setAccessible(true);
+      try {
+        result.add((String) f.get(null));
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Part of Create operation. All tmp tables are not transactional and are marked as
+   * compaction tables. Additionally...
+   * - Insert-only compaction tables copy source tables' tblproperties, except metastore/statistics
+   * properties.
+   */
+  protected void addTblProperties(StringBuilder query) {
+    Map<String, String> tblProperties = new HashMap<>();
+    tblProperties.put("transactional", "false");
+    if (sourceTab != null) { // to avoid NPEs, skip this part if sourceTab is null
+      // Exclude all standard table properties.
+      Set<String> excludes = getHiveMetastoreConstants();
+      excludes.addAll(StatsSetupConst.TABLE_PARAMS_STATS_KEYS);
+      for (Map.Entry<String, String> e : sourceTab.getParameters().entrySet()) {
+        if (e.getValue() == null) {
+          continue;
+        }
+        if (excludes.contains(e.getKey())) {
+          continue;
+        }
+        tblProperties.put(e.getKey(), HiveStringUtils.escapeHiveCommand(e.getValue()));
+      }
+    }
+    super.addTblProperties(query, tblProperties);
+  }
+
+  /**
+   * Part of Create operation. Insert-only compaction tables copy source tables' serde.
+   */
+  protected void copySerdeFromSourceTable(StringBuilder query) {
+    if (storageDescriptor != null) {
+      ensureTableToCompactIsNative();
+      SerDeInfo serdeInfo = storageDescriptor.getSerdeInfo();
+      Map<String, String> serdeParams = serdeInfo.getParameters();
+      query.append(" ROW FORMAT SERDE '").append(HiveStringUtils.escapeHiveCommand(serdeInfo.getSerializationLib()))
+          .append("'");
+      // WITH SERDEPROPERTIES
+      if (!serdeParams.isEmpty()) {
+        DDLPlanUtils.appendSerdeParams(query, serdeParams);
+      }
+      query.append("STORED AS INPUTFORMAT '")
+          .append(HiveStringUtils.escapeHiveCommand(storageDescriptor.getInputFormat())).append("'")
+          .append(" OUTPUTFORMAT '").append(HiveStringUtils.escapeHiveCommand(storageDescriptor.getOutputFormat()))
+          .append("'");
+    }
+
+  }
+
+  private void ensureTableToCompactIsNative() {
+    if (sourceTab != null) {
+      String storageHandler = sourceTab.getParameters().get(hive_metastoreConstants.META_TABLE_STORAGE);
+      if (storageHandler != null) {
+        String message =
+            "Table " + sourceTab.getTableName() + "has a storage handler (" + storageHandler + "). Failing compaction for this non-native table.";
```

Comment:
There is a space missing before has, else if table names is tb123 it will print "tbl123has" a storage handler